css/parser/tests/fixture.rs

```diff
@@ -26,9 +25,15 @@ impl Visit for AssertValid {
         }
 
         let mut errors = vec![];
+        let mut tokens = s.clone().args;
+
+        tokens.tokens.push(TokenAndSpan {
+            span: DUMMY_SP,
+            token: Token::EOF,
+        });
 
```

Comment:
This seems like a severe regression of API.

Reply:
Ideally we should never faced with it in real code, `parse_tokens` should always ending with `EOF` token, this change only because we can't parse `:pseudo-selector(content)` due broken selector parser, I think we will fix it in near future, now it is just workaround for our tests.

Reply:
Shorty - just put `EOF` token at the end if you need to parse using tokens, nothing more

Reply:
@kdy1 Friendly ping, any thoughts?

Reply:
I'm still against this PR, as it adds lots of `invalid state`s.

Reply:
@kdy1 But it helps to error reporting and recovery mode, for example if we have not closed string we will report about unexpected EOF token, but now we just report error (like expected value, but it is wrong, value exist, but EOF token unexpected here), without showing why (you can see updated snapshots here).

To be honestly when we handle tokens we have two situations:
1. broken tokens input (no EOF token at the end), if you handle tokens in stream (it very important) we can't rely on end of array (i.e. `None`), we can only understand end of file by presence EOF token
2. broken tokens input (wrong token order or unexpected token after other token), in this case we use error recovery and work as browser (i.e. skip invalid properties/values/declarations/rules/etc)

That is why almost everyone lexer has `EOF` token, many of them are written with the possibility with stream reading and just full file consumption. 

I think that in the future we may well have stream API, it can potentially increase performance in many places. Sometimes we do not need to read the entire file at once, but consumption occurs chunks.

I understand that this may look a little strange and look like `invalid state`s. But I can provide links/code of many lexers and you will find it everywhere, because as I said earlier, lexers are created no matter how the file is read in stream or buffer of the whole file or something other. 

Prof:
1. babel - https://github.com/babel/babel/blob/main/packages/babel-parser/src/tokenizer/state.js#L108
2. esbuild - https://github.com/evanw/esbuild/blob/master/internal/css_lexer/css_lexer.go#L257
3. css-tree - https://github.com/csstree/csstree/blob/a8e4f1b071a3bc2117b83eae8c3cc2aebc176405/lib/tokenizer/types.js#L3 

I can provide more if you need